Migrate stock_available_sale to v8

clonedagain opened the pull request:
- [x] Please review/merge this before: https://github.com/OCA/stock-logistics-warehouse/pull/100

Correctly segregates templates and variants.
Rewritten in new API.
Uses the ORM instead of SQL calls.
Conforms better to OCA standards

You may be interested in this other PR too: https://github.com/OCA/stock-logistics-warehouse/pull/93

Commit:
Migrate stock_available_sale to v8

Correctly segregates templates and variants.
Rewritten in new API.
Uses the ORM instead of SQL calls.
Conforms better to OCA standards
'Quoted qty' is now positive. Explain UoM feature drop.
Added directions to block quotations using sale_exceptions
Rewrote the tests in python and added new tests regarding warehouses, locations, variants and templates ; and now test the quantity available for sale.
Don't use SUPERUSER privileges anymore (it's not needed, it prevents cache reuse, it may breaks multicompany features)
Simplify copyright headers

clonedagain commented:
@lbellier for your information

clonedagain commented:
WIP, please bear with me as I iron out the details.

clonedagain commented:
The remaining failures seem unrelated to these changes, so I'd say this is ready for review.

moylop260 commented on stock_available_sale/views/product_template_view.xml:
```diff
@@ -3,8 +3,8 @@
-            <field name="name">product.form.quoted_qty</field>
-            <field name="model">product.product</field>
+            <field name="name">Quoted quantity on product form</field>
+            <field name="model">product.template</field>
```
Why not `product.product`?

clonedagain replied:
If I'm not mistaken, this view is reused for both product templates and variants so doing it on product.template serves both purposes at once.
The tree view on the other hand seems to have been specifically inherited for product.product in the standard, but I'm not adding the new field there.

moylop260 commented on stock_available_sale/models/product_template.py:
```diff
@@ -0,0 +1,47 @@
+class ProductTemplate(models.Model):
+    _inherit = 'product.template'
+
+    quoted_qty = fields.Float(
```
Why `product.template`?

clonedagain replied:
We need both: product.template shows the total of the variants. That's what the standard does with quantities too if I read it correctly.
The one for product.product is at stock_available_sale/models/product_product.py, line 27.

moylop260 replied:
Sorry, I don't understand why we need the same field in both models.
If `product_template.quoted_qty` show total of the variants then IMHO should be called different to  `product_prodcut.quoted_qty` because first one will be overwritten.

clonedagain replied:
Yes, the method of product.product is overriding product.template on purpose, similar to what the standard does: check out https://github.com/odoo/odoo/blob/8.0/addons/stock/product.py#L345 vs https://github.com/odoo/odoo/blob/8.0/addons/stock/product.py#L139

moylop260 commented:
@clonedagain 
travis red. Is a know error?

clonedagain commented:
@moylop260 The error seems unrelated but please double-check if you can.
The main 8.0 is red too, but I don't know if someone is working to make it green.

clonedagain commented:
@moylop260 if you care about this PR you may want to test #80 too, which makes the product/variant situation more correct.

gurneyalex commented:
@clonedagain build should be fixed by #82

gurneyalex commented:
:+1: 

if you have the opportunity to add an automated test using the location / warehouse context it would be a very nice addition.

gurneyalex commented:
@clonedagain can you rebase this PR and resolve the merge conflict?

clonedagain commented:
@gurneyalex I merged the latest 8.0 branch, I hope I didn't break anything.

clonedagain commented:
@gurneyalex I've improved the tests, hopefully now it's going to get the coverall quota.

clonedagain commented:
I've also removed the `@api.one` decorators like was done recently in `stock_available_immediately`.

Commit:
[IMP] Decouple the quantity for templates and variants

There are cases where we dot NOT want to simply sum the quantities of all the variants. For example when dealing with manufacturing capacities, we may have to chose between variants because we can't make ALL of them with the same components.
So instead of a simple non-modular implementation, we'll let each module define his own implementation of how to compute the product template's quantity available for sale.

clonedagain commented:
I had to rebase this on another PR: https://github.com/OCA/stock-logistics-warehouse/pull/100
This is not needed for this PR by itself, but I fear that the changes demanded by #100 will be forgotten if I don't include them right now.

Anyway, if runbot & travis are still green, this still needs review.

Commit:
Simplify copyright headers

clonedagain commented:
Travis is red on a Transifex error. Anything I can do about it?

pedrobaeza commented:
Retry as you have done.

clonedagain commented on stock_available_sale/models/product_product.py:
```diff
@@ -0,0 +1,118 @@
+            # Compute the quoted quantity in the product's UoM
+            # Rounding is OK since small values have not been squashed before
+            results += Counter({
+                product_id: self.env['product.uom'].sudo()._compute_qty(
```
@lmignon I'm doing a `sudo` also in `stock_available_sale`, do you think it could break multi-company setups like it did in `stock_available_mrp`?

lmignon replied:
@clonedagain Why do you need to use `sudo()` here? `product.uom` can be read by `base.group_user`. Moreover, each time you use `sudo()`, you build a new empty environment and the _compute_qty will need to query the database to retrieve the product, the uom, etc... IMO, you introduce here a performance issue. What do you think?

lmignon commented on stock_available_sale/models/product_product.py:
```diff
@@ -0,0 +1,118 @@
+
+        domain = [
+            ('state', '=', 'draft'),
+            ('product_id', 'in', [p.id for p in self])]
```
self._ids

clonedagain commented:
@lmignon I changed as you suggested.

lmignon commented:
:+1:  (Code review) Tahnk you @clonedagain

moylop260 commented:
:+1: (w/o test)
